NewHaven/src/Setting.cpp

```diff
@@ -0,0 +1,188 @@
+//
+// Created by hyperion on 2020-03-17.
+//
+#include "Setting.h"
+#include "GBMapLoader.h"
+#include "VGMapLoader.h"
+#include "../Exceptions/InvalidConfigurationException.h"
+#include "../Exceptions/BoardConfigurationNotLoaded.h"
+#include "iostream"
+
+Setting* Setting::current;
+
+Setting::Setting(){
+    h_deck = nullptr;
+    b_deck = nullptr;
+    board = nullptr;
+    players = nullptr;
+    current = this;
+}
+
+Setting::Setting(const Setting& setting):h_deck{new HarvestDeck(*setting.h_deck)}, b_deck{new BuildingDeck(*setting.b_deck)},
+    board{new GBMap(*setting.board)}, players{new vector<Player>(*setting.players)}{
+    // singleton design we dont need any other reference but the current one
+    delete current;
+    current = this;
+}
+
+Setting& Setting::operator=(const Setting& setting){
+    if(this == &setting)
+        return *this;
+
+    // cannot assume that the value ares nullptr and cant use assignment operator for Harvest/Building deck since implicitly deleted
+   delete h_deck;
+   delete b_deck;
+   delete board;
+   delete players;
```

Comment:
Why can't you assume all of these have been deleted already since another Setting object was obviously created?

Reply:
This is no longer relevant. It used to apply when the fields were static. I forgot to remove it when I switched the whole class into a singleton

Reply:
right no i remember now, everything defaults to a nullptr. There is no guarantee that the field is not null or does not already point to another pointer. If it does point to another data it should be deleted. I have to do this and assign it a new pointer to a copy of the data.

Reply:
Cause you need to remember that now the object is a singleton. The only static member variable is current which refers to the current active instance of the class. However, this doesn't prevent other instances of the class to exist in memory with their own member variables stored in heap. In which case you could assign one to the other making it the active instance.

Reply:
I suppose that by adding a delete current to the default constructor it should avoid this issue, and we could remove the delete from the assignment operator.